### PR TITLE
Add as65 assembler, improved errors + .BYTE support

### PR DESCRIPTION
**Pull request by rib**
Ideally this could have been split into a few patches, but still I figured I'd share some of the changes I recently made in case they are useful to others...

The changes this patch makes are:
1. Add a standalone `ca65` assembler tool
2. Switches to parsing input as a &str instead of &[u8] so it's possible to use nom's `convert_error` utility for formatting `VerboseError`s
3. Adds a custom `AssembleError` (using `thiserror` crate) that can explicitly report errors with tokenizing, parsing or IO (instead of just returning a String)
4. Uses `VerboseError` for tracking nom errors with context, to get more insightful error messages
5. Adds support for `;` based comments
6. Starts adding support for compiler directives, with support for a `.BYTE` / `.DB` directive that can be used to insert raw bytes into the assembly

_Note: The parser now only looks for / allows a single instruction per-line
Note: This removes support for labels that don't have a colon_

### CODE REVIEW

**Comment by bgourlie** [comment]
Looks good to me, but want to ping @myxo for review as well since he's been contributing lately.

**Comment by myxo** [comment]
I'll review it on the weekend

File: Cargo.toml [hunk 1]
```diff
@@ -14,3 +14,14 @@ travis-ci = { repository = "bgourlie/asm6502", branch = "master" }
...
+thiserror = "1"
+clap = { version = "3.2.6", features = ["derive", "cargo"], optional = true }
+anyhow = { version = "1.0.58", optional = true }
+
+[features]
+default=["assembler"]
+assembler=["clap", "anyhow"]
+
+[[bin]]
+name="as65"
```
Comment (myxo) [inline Cargo.toml:26]: I like the name =)

File: src/parser/mod.rs [hunk 2]
```diff
@@ -2,43 +2,87 @@
...
-pub fn parse_lines(input: &[u8]) -> IResult<&[u8], Vec<Token>> {
-    let parser = all_consuming(tuple((
-        multispace0,
-        separated_list0(multispace1, alt((opcode, label))),
-        multispace0,
-    )));
-    map(parser, |(_, ops, _)| ops)(input)
+pub type Result<'a, T> = IResult<&'a str, T, VerboseError<&'a str>>;
+
+pub fn end_of_line_comment(input: &str) -> Result<Option<&str>>
```
Comment (myxo) [inline src/parser/mod.rs]: Can you add a unit test for comments?

File: src/parser/mod.rs [hunk 3]
```diff
@@ -2,43 +2,87 @@
...
+{
```
Comment (myxo) [inline src/parser/mod.rs]: Code style is a little different. I would suggest to run `cargo fmt`, maybe as separate commit

File: src/parser/mod.rs [hunk 4]
```diff
@@ -2,43 +2,87 @@
...
+fn expression_sequence(input: &str) -> Result<Token> {
+    let (rest, bytes) = separated_list1(spaced_comma, alt((parse_signed_byte_hex, parse_signed_byte_dec)))(input)?;
+    Ok((rest, Token::ControlCommand(ControlCommand::Byte(bytes))))
+}
+
+fn byte_cmd(input: &str) -> Result<Token> {
+    let (rest, _) = alt((tag_no_case(".BYTE"), tag_no_case(".DB")))(input)?;
+    let (rest, _) = space1(rest)?;
+    expression_sequence(rest)
+}
```
Comment (myxo) [inline src/parser/mod.rs]: Also, can you add unit tests for .BYTE and .DB?

File: src/assembler/mod.rs [hunk 5]
```diff
@@ -59,490 +70,501 @@ fn process_tokens<W: Write>(tokens: Vec<Token>, output: &mut W) -> AssembleResul
...
 
 #[inline]
-fn byte<T: Write>(val: u8, output: &mut T) -> AssembleResult {
+fn byte<'a, 'b, T: Write>(_input: &'a str, val: u8, output: &'b mut T) -> AssembleResult<'a> {
     output
         .write(&[val])
         .map(|_| ())
-        .map_err(|_| "An error occurred while writing to the buffer".to_owned())
+        .map_err(|err| AssembleError::WriteIO(err))
 }
```
Comment (myxo) [inline src/assembler/mod.rs:797]: Why you pass `input` everywhere if you don't use it?

File: src/parser/tests.rs [hunk 6]
```diff
@@ -310,78 +310,68 @@ fn parse_lines() {
...
+    let parsed = super::parse_lines("label:");
     assert_eq!(
         parsed,
-        Ok((&[] as &[u8], vec![Token::Label("label".to_string())]))
-    );
-    assert!(parsed.is_ok());
-}
-
-#[test]
-fn single_label_no_colon() {
```
Comment (myxo) [inline src/parser/tests.rs:369]: Do not delete this test, just `#[ignore]` it. Maybe I'll return to it at some point and make it work.

**Review by myxo — CHANGES_REQUESTED** [review]
Basically I want unit tests for new parsed objects and clarification about `input` var in `assembler/mod.rs`
Otherwise it looks great. New error messages are helpful.

File: src/assembler/mod.rs (hunk 5 above)
Comment (rib) [reply src/assembler/mod.rs:797]: I think the error needed the lifetime of the input, so it was a way of satisfying that

**Comment by rib** [comment]
Okey, the updated branch hopefully addresses all the above comments. It adds unit tests for comments + the .byte/.db commands, fixes the braces for end_of_line_comment and also runs `cargo fmt` in a follow up patch.

After adding unit tests for comments I also ended up updating the support to handle lines that only have a comment (previously it was only handling comments after an instruction or label)

About passing `input` around; the commit message has a note about this:
> For the sake of passing the required lifetime needed for
> VerboseErrors then parsing code is now also given a reference
> to the original input.

**Comment by rib** [comment]
(oh, and I re-instated the `single_label_no_colon` test with an `#[ignore]` directive)

File: src/parser/tests.rs [hunk 7]
```diff
@@ -406,17 +396,88 @@ fn label_no_colon_in_context() {
...
+            ]
+        ))
+    );
+}
+
+#[test]
+fn parse_direct_bytes_db_alias() {
+    let asm = ".BYTE $4E, $45, $53, $1A";
+    let parsed_bytes = super::parse_lines(&asm);
+    let parsed_db = super::parse_lines(&asm);
```
Comment (myxo) [inline src/parser/tests.rs]: I'll assume this was your intension 
```suggestion
    let asm_byte = ".BYTE $4E, $45, $53, $1A";
    let asm_db = ".DB $4E, $45, $53, $1A";
    let parsed_bytes = super::parse_lines(&asm_byte);
    let parsed_db = super::parse_lines(&asm_db);
```

**Review by myxo — APPROVED** [review]
Ok, LGTM.
Thanks for PR!

**Pull request merged**